src/Symfony/Component/Message/DependencyInjection/MessagePass.php

```diff
@@ -0,0 +1,99 @@
+use Symfony\Component\DependencyInjection\ContainerBuilder;
+use Symfony\Component\DependencyInjection\Exception\RuntimeException;
+use Symfony\Component\DependencyInjection\Reference;
+
+/**
+ * @author Samuel Roze <samuel.roze@gmail.com>
+ */
+class MessagePass implements CompilerPassInterface
+{
+    use PriorityTaggedServiceTrait;
+
+    private $messageHandlerResolverService;
+    private $handlerTag;
+    private $messageBusService;
+    private $middlewareTag;
+
+    public function __construct(string $messageBusService = 'message_bus', string $middlewareTag = 'message_middleware', string $messageHandlerResolverService = 'message.handler_resolver', string $handlerTag = 'message_handler')
+    {
+        $this->messageHandlerResolverService = $messageHandlerResolverService;
+        $this->handlerTag = $handlerTag;
+        $this->messageBusService = $messageBusService;
+        $this->middlewareTag = $middlewareTag;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function process(ContainerBuilder $container)
+    {
+        if (!$container->hasDefinition($this->messageBusService)) {
+            return;
+        }
+
+        if (!$middlewares = $this->findAndSortTaggedServices($this->middlewareTag, $container)) {
+            throw new RuntimeException(sprintf('You must tag at least one service as "%s" to use the "%s" service.', $this->middlewareTag, $this->messageBusService));
+        }
+
+        $busDefinition = $container->getDefinition($this->messageBusService);
+        $busDefinition->replaceArgument(0, $middlewares);
+
+        $handlerResolver = $container->getDefinition($this->messageHandlerResolverService);
+        $handlerResolver->replaceArgument(0, $this->findHandlers($container));
+    }
+
+    private function findHandlers(ContainerBuilder $container): array
+    {
+        $handlersByMessage = array();
+
+        foreach ($container->findTaggedServiceIds($this->handlerTag, true) as $serviceId => $tags) {
+            foreach ($tags as $tag) {
+                $reflection = new \ReflectionClass($container->getDefinition($serviceId)->getClass());
+
+                try {
+                    $method = $reflection->getMethod('__invoke');
+                } catch (\ReflectionException $e) {
+                    throw new RuntimeException(sprintf('Service "%s" should have an `__invoke` function', $serviceId));
+                }
+
+                $parameters = $method->getParameters();
+                if (1 !== count($parameters)) {
+                    throw new RuntimeException(sprintf('`__invoke` function of service "%s" must have exactly one parameter', $serviceId));
+                }
+
+                $parameter = $parameters[0];
+                if (null === $parameter->getClass()) {
+                    throw new RuntimeException(sprintf('The parameter of `__invoke` function of service "%s" must type hint the Message class it handles', $serviceId));
```

Comment:
this is not ideal for generic handlers right? or with 7.2 object type. Should there be a `handles` attribute?

Reply:
I'd argue that this is rare. and that we should stand firm that the correct way is one Message should be handled by one and only one handler.

Reply:
ok. what about interface bound messages?

Reply:
I disagree here actually. It should not be a requirement, will change this.